File: content/snippets/scala2/interview_lcci_17_17_multi_search.scala
import scala.collection.mutable

object Solution {
  class TrieNode {
    val children: Array[TrieNode] = new Array[TrieNode](26)
    var idx: Int = -1
  }

  def multiSearch(big: String, smalls: Array[String]): Array[Array[Int]] = {
    val root = new TrieNode()
    for ((s, i) <- smalls.zipWithIndex if s.nonEmpty) {
      var node = root
      for (c <- s) {
        val j = c - 'a'
        if (node.children(j) == null) node.children(j) = new TrieNode()
        node = node.children(j)
      }
      node.idx = i
    }
    val ans = Array.fill(smalls.length)(mutable.ArrayBuffer[Int]())
    for (i <- big.indices) {
      var node = root
      var j = i
      var stop = false
      while (j < big.length && !stop) {
        val k = big(j) - 'a'
        if (node.children(k) == null) stop = true
        else {
          node = node.children(k)
          if (node.idx != -1) ans(node.idx) += i
          j += 1
        }
      }
    }
    ans.map(_.toArray)
  }
}

File: content/snippets/go/interview_lcci_17_17_multi_search.go
package main

type Trie struct {
	children [26]*Trie
	idx      int
}

func newTrie() *Trie {
	return &Trie{idx: -1}
}

func (t *Trie) insert(word string, i int) {
	node := t
	for _, c := range word {
		j := c - 'a'
		if node.children[j] == nil {
			node.children[j] = newTrie()
		}
		node = node.children[j]
	}
	node.idx = i
}

func multiSearch(big string, smalls []string) [][]int {
	trie := newTrie()
	for i, s := range smalls {
		if s != "" {
			trie.insert(s, i)
		}
	}

	ans := make([][]int, len(smalls))
	for i := range ans {
		ans[i] = []int{}
	}

	for i := range big {
		node := trie
		for j := i; j < len(big); j++ {
			k := big[j] - 'a'
			if node.children[k] == nil {
				break
			}
			node = node.children[k]
			if node.idx != -1 {
				ans[node.idx] = append(ans[node.idx], i)
			}
		}
	}
	return ans
}
